File: LeetCode刷题/62.不同路径.py
#
# @lc app=leetcode.cn id=62 lang=python3
#
# [62] 不同路径
#

# @lc code=start
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # 某个点的路径个数和左边的和上面的相关，找到状态转移方程
        # 处理第一行
        dp = [1] * n
        for i in range(1, m):
            tmp = [1] * n  # 每行的第 1 列一定只有 1 条路径，直着下来
            for j in range(1, n):
                tmp[j] = tmp[j-1]+dp[j]
            dp = tmp
        return dp[n-1]
# @lc code=end

File: leetcode/cpp-template/leetcode/editor/cn/62.不同路径.cpp
/*
 * @lc app=leetcode.cn id=62 lang=cpp
 *
 * [62] 不同路径
 */

#include "../common/globalmacros.h"
#include "../common/ListNode.h"
#include "../common/TreeNode.h"

// @lc code=start




class Solution
{

public:

    int uniquePaths(int m, int n)
    {
        std::vector<int> dp(n);
        // 将第一行纳入统一迭代推导出的合理的初始条件（相当于添加 -1 行）
        dp[0] = 1;

        for (int i = 1; i <= m; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                if (j) dp[j] += dp[j - 1];
            }
        }

        return dp[n - 1];
    }
};

// @lc code=end


int main()
{
    Solution solution;
    // your test code here
}
